src/java/org/apache/cassandra/streaming/StreamSession.java

```diff
@@ -661,7 +718,34 @@ public void received(TableId tableId, int sequenceNumber)
      */
     public synchronized void complete()
     {
-        logger.debug("handling Complete message, state = {}, completeSent = {}", state, completeSent);
+        logger.debug("[Stream #{}] handling Complete message, state = {}, completeSent = {}", planId(), state, completeSent);
+        if (state == State.WAIT_COMPLETE)
+        {
+            if (!completeSent)
+            {
+                messageSender.sendMessage(new CompleteMessage());
+                completeSent = true;
+            }
+            closeSession(State.COMPLETE);
+        }
+        else
+        {
+            state(State.WAIT_COMPLETE);
+        }
+    }
+
+    /**
+     * Synchronize both {@link #complete()} and {@link #maybeCompleted()} to avoid racing
+     */
+    private synchronized boolean maybeCompleted()
+    {
+        if (!(receivers.isEmpty() && transfers.isEmpty()))
+            return false;
+
+        // if already executed once, skip it
+        if (!maybeCompleted.compareAndSet(false, true))
```

Comment:
As `maybCompleted` is only used within this synchroznized method there is no need to use an `AtomicBoolean`

Reply:
+1